File: bench/polyglot/go/map_words.go
package main

import (
	"fmt"
	"strconv"
)

func mainWorkload() int64 {
	m := make(map[string]int64)
	var i int64
	for i < 30000 {
		key := "w" + strconv.FormatInt((i*i+i/3)%2000, 10)
		if _, ok := m[key]; ok {
			m[key] = m[key] + 1
		} else {
			m[key] = 1
		}
		i++
	}
	var checksum int64
	for k, v := range m {
		checksum += v * int64(len(k))
	}
	var hits int64
	var j int64
	for j < 10000 {
		probe := "w" + strconv.FormatInt((j*7)%3000, 10)
		if value, ok := m[probe]; ok {
			hits += value
		}
		j++
	}
	return checksum + hits + int64(len(m))
}

func main() {
	fmt.Println(mainWorkload())
}

File: bench/polyglot/rust/map_words.rs
use std::collections::HashMap;

fn main_workload() -> i64 {
    let mut m: HashMap<String, i64> = HashMap::new();
    let mut i: i64 = 0;
    while i < 30000 {
        let key = format!("w{}", (i * i + i / 3) % 2000);
        if m.contains_key(&key) {
            *m.get_mut(&key).expect("key present") += 1;
        } else {
            m.insert(key, 1);
        }
        i += 1;
    }
    let mut checksum: i64 = 0;
    for (k, v) in &m {
        checksum += v * k.len() as i64;
    }
    let mut hits: i64 = 0;
    let mut j: i64 = 0;
    while j < 10000 {
        let probe = format!("w{}", (j * 7) % 3000);
        if m.contains_key(&probe) {
            hits += m[&probe];
        }
        j += 1;
    }
    checksum + hits + m.len() as i64
}

fn main() {
    println!("{}", main_workload());
}
